hs_core/views/__init__.py

```diff
@@ -382,6 +382,82 @@ def unshare_resource_with_user(request, shortkey, user_id, *args, **kwargs):
+    label_type: type of label - one of these: 'LABEL', 'FAVORITE', 'MINE', 'FOLDER', 'SAVEDLABEL'
+    label: a string value required if the label_type is either 'LABEL', 'SAVEDLABEL',
+           or 'FOLDER' and action is 'CREATE'
+
+    """
+    _LABEL = 'LABEL'
+    _FOLDER = 'FOLDER'
+    _FAVORITE = 'FAVORITE'
+    _MINE = 'MINE'
+    _SAVEDLABEL = 'SAVEDLABEL'
+
+    # TODO: clear all labels, clear all saved labels
+    res, _, user = authorize(request, shortkey, view=True, full=True, superuser=True)
+    action = request.POST['action']
+    label_type = request.POST['label_type']
+    label = request.POST.get('label', None)
+    err_msg = None
+
+    # validate post data
+    if label_type not in (_LABEL, _FAVORITE, _FOLDER, _MINE, _SAVEDLABEL):
+        err_msg = "Invalid type of label found"
+    elif label_type == _LABEL or label_type == _SAVEDLABEL:
+        if label is None:
+            err_msg = "A value for label is missing"
+    elif label_type == _FOLDER and action == 'CREATE':
+        if label is None:
+            err_msg = "A folder name is missing"
+
+    if err_msg is None:
+        try:
+            if action == 'CREATE':
+                if label_type == _LABEL:
+                    hydroshare.label_resource(user, res, label)
+                elif label_type == _SAVEDLABEL:
+                    hydroshare.create_resource_label(user, label)
+                elif label_type == _FAVORITE:
+                    hydroshare.favorite_resource(user, res)
+                elif label_type == _FOLDER:
+                    hydroshare.create_resource_folder(user, res, label)
+                elif label_type == _MINE:
+                    hydroshare.add_to_my_resources(user, res)
+            elif action == 'DELETE':
+                if label_type == _LABEL:
+                    hydroshare.unlabel_resource(user, res, label)
+                elif label_type == _SAVEDLABEL:
+                    hydroshare.delete_resource_label(user, label)
+                elif label_type == _FAVORITE:
+                    hydroshare.unfavorite_resource(user, res)
+                elif label_type == _FOLDER:
+                    hydroshare.delete_resource_folder(user, res)
+                elif label_type == _MINE:
+                    hydroshare.delete_from_my_resources(user, res)
+
+        except Exception as exp:
+            err_msg = exp.message
+
+    if request.is_ajax():
+        if err_msg:
+            ajax_response_data = {'status': 'error', 'message': err_msg}
+        else:
+            ajax_response_data = {'status': 'success', 'label_type': label_type, 'action': action,
+                                  'resource_id': shortkey}
+
+        return HttpResponse(json.dumps(ajax_response_data))
+
+    if err_msg:
+        messages.error(request, err_msg)
+    else:
+        messages.success(request, "Resource labeling action was successful")
+    return HttpResponseRedirect(request.META['HTTP_REFERER'])
```

Comment:
Checked with @mauriel133 and he says he needs only AJAX support. Will make the change.